auth/src/main/java/com/firebase/ui/auth/ui/email/RegisterEmailFragment.java

```diff
@@ -297,12 +300,63 @@ public void onFailure(@NonNull Exception e) {
                         } else if (e instanceof FirebaseAuthUserCollisionException) {
                             // Collision with existing user email, it should be very hard for
                             // the user to even get to this error due to CheckEmailFragment.
-                            mEmailInput.setError(getString(R.string.error_user_collision));
+
+                            ProviderUtils.fetchTopProvider(mHelper.getFirebaseAuth(), email).addOnSuccessListener(
+                                    getActivity(),
+                                    new OnSuccessListener<String>() {
+                                        @Override
+                                        public void onSuccess(String provider) {
+                                            Toast.makeText(getContext(),
+                                                           R.string.error_user_collision,
+                                                           Toast.LENGTH_LONG)
+                                                    .show();
+
+                                            if (provider == null) {
+                                                throw new IllegalStateException(
+                                                        "User has no providers even though we got a FirebaseAuthUserCollisionException");
+                                            } else if (EmailAuthProvider.PROVIDER_ID.equalsIgnoreCase(
+                                                    provider)) {
+                                                getActivity().startActivityForResult(
+                                                        WelcomeBackPasswordPrompt.createIntent(
+                                                                getContext(),
+                                                                mHelper.getFlowParams(),
+                                                                new IdpResponse.Builder(
+                                                                        EmailAuthProvider.PROVIDER_ID,
+                                                                        email).build()),
+                                                        RegisterEmailActivity.RC_WELCOME_BACK_IDP);
+                                            } else {
+                                                onExistingIdpUser(new User.Builder(email)
+                                                                          .setProvider(provider)
+                                                                          .build());
+                                            }
+                                        }
+
+                                        private void onExistingIdpUser(User user) {
```

Comment:
Let's not add private methods to listeners.  It's hard to follow and makes them non-lambda-compatible in the future.